File: clients/ios/Whisp/Whisp/WebSocketClient.swift
import Foundation
import Combine

class WebSocketClient: NSObject, ObservableObject {
    static let shared = WebSocketClient()
    
    private var webSocketTask: URLSessionWebSocketTask?
    private var urlSession: URLSession?
    private let baseURL: String
    
    private override init() {
        // Get WebSocket base URL from build configuration
        #if DEBUG
        // Development environment
        if let configURL = Bundle.main.object(forInfoDictionaryKey: "WS_BASE_URL") as? String {
            self.baseURL = configURL
        } else {
            self.baseURL = "ws://localhost:4000"
        }
        #else
        // Production environment
        if let configURL = Bundle.main.object(forInfoDictionaryKey: "WS_BASE_URL") as? String {
            self.baseURL = configURL
        } else {
            self.baseURL = "wss://api.whisp.app"
        }
        #endif
        super.init()
        setupMessageHandling()
    }
    
    @Published var isConnected = false
    @Published var messages: [Message] = []
    
    private var messageSubject = PassthroughSubject<WebSocketMessage, Never>()
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Connection Management
    func connect(token: String) {
        guard let url = URL(string: "\(baseURL)/ws?token=\(token)") else {
            print("Invalid WebSocket URL")
            return
        }
        
        let config = URLSessionConfiguration.default
        urlSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
        
        webSocketTask = urlSession?.webSocketTask(with: url)
        webSocketTask?.resume()
        
        receiveMessage()
    }
    
    func disconnect() {
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        urlSession = nil
        isConnected = false
    }
    
    // MARK: - Message Sending
    func sendChatMessage(to username: String, text: String) {
        let message = ChatMessage(type: "chat", to: username, text: text)
        sendMessage(message)
    }
    
    func sendFileMessage(to username: String, name: String, mime: String, size: Int, dataB64: String) {
        let message = FileMessage(type: "file", to: username, name: name, mime: mime, size: size, dataB64: dataB64)
        sendMessage(message)
    }
    
    private func sendMessage<T: Codable>(_ message: T) {
        guard let webSocketTask = webSocketTask else { return }
        
        do {
            let data = try JSONEncoder().encode(message)
            let message = URLSessionWebSocketTask.Message.data(data)
            webSocketTask.send(message) { error in
                if let error = error {
                    print("Failed to send message: \(error)")
                }
            }
        } catch {
            print("Failed to encode message: \(error)")
        }
    }
    
    // MARK: - Message Receiving
    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .data(let data):
                    self?.handleDataMessage(data)
                case .string(let string):
                    if let data = string.data(using: .utf8) {
                        self?.handleDataMessage(data)
                    }
                @unknown default:
                    break
                }
                self?.receiveMessage() // Continue receiving
            case .failure(let error):
                print("WebSocket receive error: \(error)")
                DispatchQueue.main.async {
                    self?.isConnected = false
                }
            }
        }
    }
    
    private func handleDataMessage(_ data: Data) {
        do {
            let message = try JSONDecoder().decode(WebSocketMessage.self, from: data)
            DispatchQueue.main.async {
                self.messageSubject.send(message)
            }
        } catch {
            print("Failed to decode WebSocket message: \(error)")
        }
    }
    
    // MARK: - Message Handling
    private func setupMessageHandling() {
        messageSubject
            .sink { [weak self] message in
                self?.processMessage(message)
            }
            .store(in: &cancellables)
    }
    
    private func processMessage(_ message: WebSocketMessage) {
        switch message.type {
        case "presence":
            // Handle presence updates if needed
            break
        case "chat":
            if let from = message.from,
               let text = message.text,
               let timestamp = message.timestamp {
                let newMessage = Message(
                    from: from,
                    text: text,
                    timestamp: timestamp,
                    type: .chat
                )
                messages.append(newMessage)
            }
        case "file":
            if let from = message.from,
               let name = message.name,
               let mime = message.mime,
               let size = message.size,
               let dataB64 = message.dataB64,
               let timestamp = message.timestamp {
                // For file messages, we'll store the text as the filename for now
                let newMessage = Message(
                    from: from,
                    text: "📎 \(name)",
                    timestamp: timestamp,
                    type: .file
                )
                messages.append(newMessage)
            }
        case "delivery":
            // Handle delivery status if needed
            break
        default:
            break
        }
    }
    
    // MARK: - Message Management
    func clearMessages() {
        messages.removeAll()
    }
    
    func getMessagesForUser(_ username: String) -> [Message] {
        return messages.filter { $0.from == username || $0.from == "me" }
    }
}

// MARK: - URLSessionWebSocketDelegate
extension WebSocketClient: URLSessionWebSocketDelegate {
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
        DispatchQueue.main.async {
            self.isConnected = true
        }
    }
    
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        DispatchQueue.main.async {
            self.isConnected = false
        }
    }
}

File: clients/android/app/src/main/java/com/rootkings/whisp/network/WebSocketClient.kt
package com.rootkings.whisp.network

import android.util.Log
import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import com.rootkings.whisp.models.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import okhttp3.*
import java.util.concurrent.TimeUnit

class WebSocketClient(private val context: android.content.Context) {
    
    private val baseUrl: String
    private var webSocket: WebSocket? = null
    private var okHttpClient: OkHttpClient? = null
    private val gson = Gson()
    
    private val _isConnected = MutableSharedFlow<Boolean>()
    val isConnected: SharedFlow<Boolean> = _isConnected.asSharedFlow()
    
    private val _messages = MutableSharedFlow<Message>()
    val messages: SharedFlow<Message> = _messages.asSharedFlow()
    
    private val messageList = mutableListOf<Message>()
    
    init {
        baseUrl = if (com.rootkings.whisp.BuildConfig.DEBUG) {
            // "ws://10.0.2.2:4000" // Android emulator localhost
            "ws://localhost:4000" // Use this to set IP of server
        } else {
            "wss://api.whisp.app"
        }
    }
    
    // MARK: - Connection Management
    fun connect(token: String) {
        disconnect() // Close existing connection
        
        okHttpClient = OkHttpClient.Builder()
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .build()
        
        val request = Request.Builder()
            .url("$baseUrl/ws?token=$token")
            .build()
        
        webSocket = okHttpClient?.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d("WebSocket", "Connected")
                CoroutineScope(Dispatchers.Main).launch {
                    _isConnected.emit(true)
                }
            }
            
            override fun onMessage(webSocket: WebSocket, text: String) {
                Log.d("WebSocket", "Message received: $text")
                handleMessage(text)
            }
            
            override fun onMessage(webSocket: WebSocket, bytes: okio.ByteString) {
                Log.d("WebSocket", "Binary message received")
                handleMessage(bytes.utf8())
            }
            
            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.d("WebSocket", "Closing: $code $reason")
                CoroutineScope(Dispatchers.Main).launch {
                    _isConnected.emit(false)
                }
            }
            
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d("WebSocket", "Closed: $code $reason")
                CoroutineScope(Dispatchers.Main).launch {
                    _isConnected.emit(false)
                }
            }
            
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e("WebSocket", "Error: ${t.message}", t)
                CoroutineScope(Dispatchers.Main).launch {
                    _isConnected.emit(false)
                }
            }
        })
    }
    
    fun disconnect() {
        webSocket?.close(1000, "Disconnecting")
        webSocket = null
        okHttpClient?.dispatcher?.executorService?.shutdown()
        okHttpClient = null
    }
    
    // MARK: - Message Sending
    fun sendChatMessage(to: String, text: String) {
        val message = ChatMessage("chat", to, text)
        sendMessage(message)
    }
    
    fun sendFileMessage(to: String, name: String, mime: String, size: Int, dataB64: String) {
        val message = FileMessage("file", to, name, mime, size, dataB64)
        sendMessage(message)
    }
    
    private fun sendMessage(message: Any) {
        try {
            val json = gson.toJson(message)
            webSocket?.send(json)
            Log.d("WebSocket", "Message sent: $json")
        } catch (e: Exception) {
            Log.e("WebSocket", "Failed to send message: ${e.message}", e)
        }
    }
    
    // MARK: - Message Handling
    private fun handleMessage(json: String) {
        try {
            val message = gson.fromJson(json, WebSocketMessage::class.java)
            processMessage(message)
        } catch (e: JsonSyntaxException) {
            Log.e("WebSocket", "Failed to parse message: $json", e)
        } catch (e: Exception) {
            Log.e("WebSocket", "Error handling message: ${e.message}", e)
        }
    }
    
    private fun processMessage(message: WebSocketMessage) {
        when (message.type) {
            "presence" -> {
                // Handle presence updates if needed
                Log.d("WebSocket", "Presence update received")
            }
            "chat" -> {
                message.from?.let { from ->
                    message.text?.let { text ->
                        message.timestamp?.let { timestamp ->
                            val newMessage = Message(
                                from = from,
                                text = text,
                                timestamp = timestamp,
                                type = Message.MessageType.CHAT
                            )
                            messageList.add(newMessage)
                            CoroutineScope(Dispatchers.Main).launch {
                                _messages.emit(newMessage)
                            }
                        }
                    }
                }
            }
            "file" -> {
                message.from?.let { from ->
                    message.name?.let { name ->
                        message.timestamp?.let { timestamp ->
                            val newMessage = Message(
                                from = from,
                                text = "📎 $name",
                                timestamp = timestamp,
                                type = Message.MessageType.FILE
                            )
                            messageList.add(newMessage)
                            CoroutineScope(Dispatchers.Main).launch {
                                _messages.emit(newMessage)
                            }
                        }
                    }
                }
            }
            "delivery" -> {
                // Handle delivery status if needed
                Log.d("WebSocket", "Delivery status received")
            }
        }
    }
    
    // MARK: - Message Management
    fun clearMessages() {
        messageList.clear()
    }
    
    fun getMessagesForUser(username: String): List<Message> {
        return messageList.filter { it.from == username || it.from == "me" }
    }
    
    fun addLocalMessage(text: String, timestamp: Long) {
        val message = Message(
            from = "me",
            text = text,
            timestamp = timestamp,
            type = Message.MessageType.CHAT
        )
        messageList.add(message)
        CoroutineScope(Dispatchers.Main).launch {
            _messages.emit(message)
        }
    }
}
